### PR TITLE
Scala 2.12 support

### PR DESCRIPTION
**Pull request by gkossakowski**
Adds support for Scala 2.12.3. In the history, you can see that I tested Scala 2.12.4 nightly builds to verify the issue with the flat classpath caching.

Notes:
  - Scala 2.12.3 has a bug related to classpath caching that breaks bazel's incremental compilation when scalac worker is enabled: https://github.com/bazelbuild/rules_scala/pull/251#issuecomment-333246940
  - this PR disables the jmh test, the rule doesn't support Java 8 bytecode, see #295

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by johnynek** [comment]
adding CLA yes since Grzeg is at Stripe and we have signed the CLA.

File: twitter_scrooge/twitter_scrooge.bzl [hunk 1]
```diff
@@ -9,7 +9,7 @@ load("//scala:scala.bzl",
 def twitter_scrooge():
   native.maven_server(
     name = "twitter_scrooge_maven_server",
-    url = "http://mirror.bazel.build/repo1.maven.org/maven2/",
+    url = "http://repo1.maven.org/maven2/",
```
Comment (ittaiz) [inline twitter_scrooge/twitter_scrooge.bzl]: Why can't we use the Bazel mirror?

File: WORKSPACE [hunk 2]
```diff
@@ -23,23 +23,23 @@ specs2_junit_repositories()
 # test adding a scala jar:
 maven_jar(
   name = "com_twitter__scalding_date",
-  artifact = scala_mvn_artifact("com.twitter:scalding-date:0.16.0-RC4"),
-  sha1 = "659eb2d42945dea37b310d96af4e12bf83f54d14"
+  artifact = scala_mvn_artifact("com.twitter:scalding-date:0.17.0"),
```
Comment (ittaiz) [inline WORKSPACE]: Can we do the version upgrades in master to minimize the diff?

File: src/java/io/bazel/rulesscala/scalac/BUILD [hunk 3]
```diff
@@ -6,6 +6,12 @@ java_binary(
         "ScalacProcessor.java",
     ],
     main_class = "io.bazel.rulesscala.scalac.ScalaCInvoker",
+    # this probably should be set globally for the entire rules_scala
```
Comment (ittaiz) [inline src/java/io/bazel/rulesscala/scalac/BUILD]: Maybe this needs to be in the scala.bzl?
If so we need to also test that the user gets a clear error message when they request a lower version. 
And also they can request a higher version and it's ok

File: WORKSPACE (hunk 2 above)
Comment (johnynek) [reply WORKSPACE]: +1

File: specs2/specs2.bzl [hunk 4]
```diff
@@ -4,36 +4,36 @@ def specs2_repositories():
 
   native.maven_jar(
       name = "io_bazel_rules_scala_org_specs2_specs2_core",
-      artifact = "org.specs2:specs2-core_2.11:" + specs2_version(),
-      sha1 = "495bed00c73483f4f5f43945fde63c615d03e637",
+      artifact = "org.specs2:specs2-core_2.12:" + specs2_version(),
```
Comment (johnynek) [inline specs2/specs2.bzl]: if we used `scala_mvn_artifact` we wouldn't have this change here.

File: scala_proto/scala_proto.bzl [hunk 5]
```diff
@@ -81,8 +81,8 @@ def scala_proto_repositories():
 
     native.maven_jar(
         name = "scala_proto_rules_scalapb_lenses",
-        artifact = "com.trueaccord.lenses:lenses_2.11:0.4.12",
-        sha1 = "c5fbf5b872ce99d9a16d3392ccc0d15a0e43d823",
+        artifact = "com.trueaccord.lenses:lenses_2.12:0.4.12",
```
Comment (johnynek) [inline scala_proto/scala_proto.bzl]: if we used `scala_mvn_artifact` here we wouldn't have this change and others in the file.

File: twitter_scrooge/twitter_scrooge.bzl (hunk 1 above)
Comment (gkossakowski) [reply twitter_scrooge/twitter_scrooge.bzl]: i can't remember the reason to switch, switched back

File: WORKSPACE (hunk 2 above)
Comment (gkossakowski) [reply WORKSPACE]: #303

File: specs2/specs2.bzl (hunk 4 above)
Comment (gkossakowski) [reply specs2/specs2.bzl]: I went wholesale on this in #303

File: src/java/io/bazel/rulesscala/scalac/BUILD (hunk 3 above)
Comment (gkossakowski) [reply src/java/io/bazel/rulesscala/scalac/BUILD]: Is there a pattern of enforcing minimal Java version on clients of the rule?

**Comment by johnynek** [comment]
I'm going to close this in favor of #304

**Pull request closed without merging**